### DISCUSSION TITLE
Tests loginUser and UserChecker

### DISCUSSION THREAD

**Discussion by PyroShape** [symfony/symfony#50464] [Q&A] [answered]
Hello,

Is there any way to use the loginUser method in tests and call the UserCheckerInterface ?
The doc : [https://symfony.com/doc/current/testing.html#logging-in-users-authentication]

When I use this method, the login works but every method inside the user checker are not called (a lot of things are set up in my app in these methods).
Do you know how to call them after the login method or maybe enabled the user checker when I call it ?

**Comment by MatTheCat** [comment]
Hi,

calling `loginUser` means “this user is authenticated”, so it is already too late to make the authentication fail.

If you want to test your authentication process, you must not use `loginUser`.

**Reply by PyroShape** [reply]
Hi,

That means that to log in the application for the tests, I need to use the crawler and the login form in order to pass through the user checker ?
It is quite time consuming.
Is there any other way ?

**Reply by MatTheCat** [reply]
As I said this depends whether you're testing your authentication or not. If yes you need to authenticate as a user would do. Else you can call `loginUser` to skip a successful authentication step.

**Reply by PyroShape** [reply]
I don't want to test the authentification, just all the rest of the application that needs to have a user login.

But when I use the loginUser, the userChecker is not called (as you well explain). But inside the userChecker, and especially in the checkPostAuth method,I setup a lot of session variable that my app needs after the login.

Is it possible to call the checkPostAuth method after the loginUser in order to setup these session variable as if the user was logged throught a normal login form ?

**Reply by MatTheCat** [reply, marked as answer]
Seems weird to me to set session variables in user checkers: these are supposed to check accounts status. I guess I would have used a [security event](https://symfony.com/doc/current/security.html#authentication-events) listener instead.

Anyways, as you can access the application's container in your test that means you should be able to call the service responsible to set these session variables yourself.
